README.md
# MMM-HabiticaStats: A MagicMirror² Module
`MMM-HabiticaStats` is a module for [MagicMirror²](https://github.com/MichMich/MagicMirror) that displays player statistics for [Habitica](https://habitica.com).

 **REQUIRED API KEYS:**

 You're gonna need a (free) Habitica account to get an API token (see the [Habitica wiki](https://habitica.fandom.com/wiki/API_Options#API_Token) for details on how to find your API Token in your Habitica account).

## Screenshots

### Example Screenshots of MMM-HabiticaStats running in the default MagicMirror² theme:

<img width="400" src="screenshots/MM_portrait_DefaultTheme.png">

-------

<img width="600" src="screenshots/MM_landscape_DefaultTheme.png">


### Example Screenshots of MMM-HabiticaStats running in the [WallberryTheme](https://github.com/delightedCrow/WallberryTheme):

<img width="600" src="screenshots/WB_landscape_fullscreen.jpg">

## Installing the Module

To use this module:
1.  Copy the `MMM-HabiticaStats` folder to your `MagicMirror/modules` directory
2.  Add the module to the modules array in the `config/config.js` file like in the following example:

````javascript
modules: [
	{
		module: "MMM-HabiticaStats",
		position: "middle_center", // put it wherever you want
		config: {
			userID: "YOUR-HABITICA-USER-ID-HERE",
			APIToken: "YOUR-HABITICA-API-TOKEN-HERE"
		}
	}
]
````

## Configuration options

The following properties can be configured:


| Option                      | Type    | Description
| ----------------------------|---------| -----------
| `userID`                    | String  | Your Habitica userID. Sign up for free at [https://habitica.com](https://habitica.com). <br><br>  This value is **REQUIRED**
| `APIToken`                  | String  | Your Habitica API Token. <br><br>  This value is **REQUIRED**
| `refreshRate`               | Number  | How often the module should refresh your Habitica data (in milliseconds). <br><br> **Possible values:** Any positive integer <br> **Default value:** `60 * 60 * 1000`, 1 hour
| `backgroundBlurOn`          | Boolean | Turn the background blur on; good for making the text more legible on photo backgrounds. <br><br> **Possible values:** `true` (turn on) or `false` (turn off) <br> **Default value:** `true`
| `zoom`                      | Float   | Scale this module using the CSS zoom property. <br><br> **Possible values:** A float number. Examples: `1.5` (scale up 150%), `0.7` (scale down 70%) <br> **Default value:** none
| `orientation`               | String  | Whether to display the module in portrait or landscape orientation. <br><br> **Possible values:** `portrait`, `landscape`<br> **Default value:** `portrait`
| `positionCSS`               | String  | Finely position this module with the power of CSS. BE CAREFUL, you can break all kinds of things with this. <br><br> **Possible values:** Any string of valid CSS. Example: `"float: right; position: relative; top: -10px;"` <br> **Default value:** none

## Fixing The Background Blur

##### Please make sure you have the config option `backgroundBlurOn: true` set first!

If you're running MagicMirror using Electron (i.e, not in `serveronly` mode) and the background blur isn't showing up for you, try enabling Electron's `experimentalFeatures` option by adding the following block to your `config` options in your MagicMirror `config.js` file:

````javascript
electronOptions: {
	webPreferences: {
		experimentalFeatures: true,
		nodeIntegration: false
	}
},
````

## Contributing & Development

### Changelog

To view the changelog see [CHANGELOG.md](CHANGELOG.md).

### Building CSS

This module uses SASS, so to make style changes please edit the `MMM-HabiticaStats.scss` file and use node-sass to compile the css file:

- Run the command `npm install` to install node-sass.
- Run `npm run watch` and node-sass will autocompile the CSS file any time changes are made to the `MMM-HabiticaStats.scss` file.
- Run `npm run build` to build the CSS file.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- Nothing yet :)

## 1.3.0 - 2020-4-6
### Added
- Added the latest MMM-HabiticaStats version number to the xclient header so Habitica can tell which versions of MMM-HabiticaStats are making which calls.

### Fixed
- Adding in a guard so that refreshRate for accessing the API cannot be set lower than 30 seconds in the config.
- Fixing error where we were using setInterval instead of setTimeout in our scheduleUpdate function, resulting in too many requests to the Habitica API over time (fixes issue #1)

## 1.2.0 - 2019-11-15
### Changed
- Updating to HabiticaMagic-v2.0.1 to fix mage class title and icons.
- Changed header icon to FontAwesome Scroll to better fit Habitica's image guidelines.
- Cleaning up screenshots.

### Added
- Now displays a "loading..." message while data is being fetched for the first time.

## 1.1.0 - 2019-11-4
### Added
- The changelog file :)
- Error handling and error display messages.

### Changed
- Updated to (HabiticaMagicJS 2.0.0)[https://github.com/delightedCrow/HabiticaMagic].

## 1.0.0 - 2019-10-29
The Initial Release.


HabiticaMagic.js
/* HabiticaMagic.js
 *
 * A convienient way to interact with the Habitica API
 * (https://habitica.com/apidoc/).
 *
 * By JSC (@delightedCrow) & PJM (@ArrayOfFrost)
 * Copyright © 2019 - MIT Licensed
 *
 */

class HabiticaUserTasksManager {
	constructor(data) {
		this.apiData = data;
	}

	get taskList() {
		return this.apiData;
	}

	get todosDueToday() {
		return this.todosDueOnDate(moment().endOf('day'));
	}

	todosDueOnDate(dueDate) {
		var todos = [];

		for (var i=0; i<this.taskList.length; i++) {
			let task = this.taskList[i];
			if (task.type != "todo") { continue; }

			if (task.date) {
				var taskTime = moment(task.date);
				if (taskTime.isBefore(dueDate)) {
						todos.push(task);
				}
			}
		}

		return todos;
	}

	calculateDailyStatsFor(user) {

		var stats = {
			dueCount: 0,
			totalDamageToSelf: 0,
			dailyDamageToSelf: 0,
			bossDamage: 0,
			dailiesEvaded: 0
		}
		const min = -47.27; // task value cap min
		const max = 21.27; // task value cap max
		var stealthRemaining = user.stealth;
		let conBonus = user.constitutionBonus;

		for (var i=0; i<this.taskList.length; i++) {
			let task = this.taskList[i];
			// skip the rest of this for loop if the task isn't a daily, isn't due, or is completed
			if (task.type != "daily") { continue; }
			if (!task.isDue || task.completed) { continue; }

			// thieves can cast stealth to avoid a certain number of dailies
			if (stealthRemaining > 0) { // avoided!
				stealthRemaining --;
				stats.dailiesEvaded ++;
			} else { // calculate damage!

				stats.dueCount ++;

				var taskDamage = (task.value < min) ? min : task.value;
				taskDamage = (taskDamage > max) ? max : taskDamage;
				taskDamage = Math.abs(Math.pow(0.9747, taskDamage));

				// if a subtask is completed, decrease the task damage proportionately
				if (task.checklist.length > 0 ) {
					var subtaskDamage = (taskDamage/task.checklist.length);
					for (var j=0; j<task.checklist.length; j++) {
						if (task.checklist[j].completed) {
							taskDamage = taskDamage - subtaskDamage;
						}
					}
				}

				var damage = taskDamage * conBonus * task.priority * 2;
				stats.dailyDamageToSelf += Math.round(damage * 10) / 10; // round damage to nearest tenth because game does

				// if we have a quest and a boss we can calculate the damage to party from this daily
				if (user.isOnBossQuest) {
					var bossDamage = (task.priority < 1) ? (taskDamage * task.priority) : taskDamage;
					bossDamage *= user.quest.data.boss.str;
					stats.bossDamage += bossDamage;
				}
			}
		}
		// formatting display - rounding up to be safe like Lady Alys does :) - see https://github.com/Alys/tools-for-habitrpg/blob/29710e0f99c9d706d6911f49d60bcceff2792768/habitrpg_user_data_display.html#L1952
		stats.totalDamageToSelf = stats.dailyDamageToSelf + stats.bossDamage;
		stats.totalDamageToSelf = Math.ceil(stats.totalDamageToSelf * 10) / 10;
		stats.bossDamage = Math.ceil(stats.bossDamage * 10) / 10;

		this.dailyStats = stats;
	}
}

class HabiticaUser {
	constructor(data) {
		this.apiData = data;
		this.stats = this._calculateStats();
	}

	get gems() {
		return this.apiData.balance * 4;
	}
	get hourglasses() {
		return this.apiData.purchased.plan.consecutive.trinkets;
	}
	get gold() {
		return Math.round(this.apiData.stats.gp);
	}
	get goldCompact() {
		const formatter = new Intl.NumberFormat('lookup', {
			notation: 'compact',
			compactDisplay: 'short',
		});
		return formatter.format(this.gold);
	}
	get level() {
		return this.apiData.stats.lvl;
	}
	get displayName() {
		return this.apiData.profile.name;
	}
	get className() {
		return this.apiData.stats.class;
	}
	get experience() {
		return Math.floor(this.apiData.stats.exp);
	}
	get experienceToLevel() {
		return Math.round(this.apiData.stats.toNextLevel);
	}
	get mana() {
		return Math.floor(this.apiData.stats.mp);
	}
	get manaMax() {
		return Math.round(this.apiData.stats.maxMP);
	}
	get health() {
		return Math.floor(this.apiData.stats.hp);
	}
	get healthMax() {
		return Math.round(this.apiData.stats.maxHealth);
	}
	get stealth() {
		return this.apiData.stats.buffs.stealth;
	}
	get armor() {
		return this.apiData.items.gear.equipped;
	}
	get costume() {
		return this.apiData.items.gear.costume;
	}
	get outfit() {
		return this.apiData.preferences.costume == true ? this.costume : this.armor;
	}
	get isSleeping() {
		return this.apiData.preferences.sleep;
	}


	_calculateStats() {
		var stats = {
			totals: {str: 0, con: 0, int: 0, per: 0},
			armor: {str: 0, con: 0, int: 0, per: 0},
			buffs: {
				str: this.apiData.stats.buffs.str,
				con: this.apiData.stats.buffs.con,
				int: this.apiData.stats.buffs.int,
				per: this.apiData.stats.buffs.per
			},
			points: {
				str: this.apiData.stats.str,
				con: this.apiData.stats.con,
				int: this.apiData.stats.int,
				per: this.apiData.stats.per
			}
		}

		// calculate armor stats from each piece of armor
		for (var key in this.armor) {
			let item = this.armor[key];
			for (var stat in stats.armor) {
				stats.armor[stat] += item[stat];
				// apply class bonus if user is wearing special class gear
				if (this.className === item.klass ||
						this.className === item.specialClass) {
					stats.armor[stat] += .5 * item[stat];
				}
			}
		}

		// add up all stats for total, including level bonus
		let levelBonus = Math.floor(this.level / 2);
		for (var stat in stats.totals) {
			stats.totals[stat] = stats.armor[stat] +
				stats.buffs[stat] +
				stats.points[stat] +
				levelBonus;
		}

		return stats;
	}

	// this user's constitution bonus against daily damage
	get constitutionBonus() {
		let bonus = 1 - (this.stats.totals.con / 250);
		return (bonus < 0.1) ? 0.1 : bonus;
	}

	get quest() {
		return (this.apiData.party.quest);
	}

	get isOnQuest() {
		if (this.quest.data != null) {
			return true;
		}
		return false;
	}

	get isOnBossQuest() {
		if (this.isOnQuest && this.quest.data.boss != null) {
			return true;
		}
		return false;
	}

	set tasks(userTaskManager) {
		this._taskManager = userTaskManager;
		this.tasks.calculateDailyStatsFor(this);
	}

	get tasks() {
		return this._taskManager;
	}
}

class HabiticaAPIManager {
	constructor(language="en", xclient) {
		this.language = language;
		// For more info on the xclient header: https://habitica.fandom.com/wiki/Guidance_for_Comrades#X-Client_Header
		this.xclient = xclient;
		this.content = {};
	}

	fetchContentData(callback) {
		const url = "https://habitica.com/api/v3/content?language=" +
			this.language;
		let req = this.constructor.APIRequest(url, this.xclient);
		let hm = this;

		req.onload = function() {
			if (this.status == 200) {
				let data = JSON.parse(this.responseText);
				hm.content = data.data;
				if (callback) {
					callback();
				}
			}
		};
		req.send();
	}

	fetchAuthenticatedUser(userID, userAPIToken, callback) {
		const url = "https://habitica.com/api/v3/user";
		let req = this.constructor.authenticatedAPIRequest(url, this.xclient, userID, userAPIToken);
		let hm = this;

		req.onload = function() {
			if (this.status == 200) {
				var data = JSON.parse(this.responseText).data;
				let user = new HabiticaUser(hm.replaceKeysWithContent(data));
				callback(user);
			}
		};
		req.send();
	}

	fetchUserTasks(userID, userAPIToken, callback) {
		const url = "https://habitica.com/api/v3/tasks/user";
		let req = this.constructor.authenticatedAPIRequest(url, this.xclient, userID, userAPIToken);
		let hm = this;

		req.onload = function() {
			if (this.status == 200) {
				var data = JSON.parse(this.responseText).data;
				var tasks = new HabiticaUserTasksManager(data);
				callback(tasks);
			}
		};
		req.send();
	}

	fetchUserWithTasks(userID, userAPIToken, callback) {
		this.fetchAuthenticatedUser(userID, userAPIToken, (user) => {
			this.fetchUserTasks(userID, userAPIToken, (tasks) => {
				user.tasks = tasks;
				callback(user);
			});
		});
	}

	replaceKeysWithContent(data) {
		// replace equipped and costume gear with full content version
		for (var section of [data.items.gear.equipped, data.items.gear.costume]) {
			for (var key in section) {
				let armorName = section[key];
				let armor = this.content.gear.flat[armorName];
				section[key] = armor;
			}
		}
		// replace party quest key with actual quest
		if (data.party.quest.key) {
			data.party.quest.data = this.content.quests[data.party.quest.key];
		}
		return data;
	}

	// API REQUEST FUNCTIONS
	static authenticatedAPIRequest(url, xclient, userID, userAPIToken) {
		let req = new XMLHttpRequest();
		req.open("GET", url);

		req.onerror = function() {
			console.error("HabiticaAPI Error: ", this.statusText);
		};

		req.setRequestHeader("x-api-user", userID);
		req.setRequestHeader("x-api-key", userAPIToken);
		req.setRequestHeader("x-client", xclient);
		return req;
	}

	static APIRequest(url, xclient) {
		let req = new XMLHttpRequest();
		req.open("GET", url);

		req.onerror = function() {
			console.error("HabiticaAPI Error: ", this.statusText);
		};

		req.setRequestHeader("x-client", this.xclient);

		return req;
	}
}


MMM-HabiticaStats.js
/* Magic Mirror
 * Module: MMM-HabiticaStats
 *
 * By JSC (@delightedCrow) & PJM (@ArrayOfFrost)
 * Copyright © 2019 - MIT Licensed
 *
 */
Module.register("MMM-HabiticaStats", {
	defaults: {
		userID: null, // required
		APIToken: null, // required

		refreshRate: 60 * 60 * 1000, // 1 hour, in milliseconds
		backgroundBlurOn: true,
		zoom: null,
		orientation: "portrait", // "portrait" or "landscape"
		positionCSS: null // put the module where you want
	},

	user: null,
	fetchError: null,
	updateTimer: null,
	minRefreshRate: 30 * 1000, // minimum Refresh rate should be 30s

	getScripts: function() {
		return ["moment.js", this.file("vendor/HabiticaMagic-v2.0.1.min.js")];
	},

	getStyles: function() {
		return [
			this.file("assets/MMM-HabiticaStats.css"),
			'font-awesome.css'
		];
	},

	getTemplate: function() {
		return "MMM-HabiticaStats.njk"
	},

	getTemplateData: function() {
		return {
			user: this.user,
			config: this.config,
			error: this.fetchError
		};
	},

	start: function() {
		Log.info("Starting module: " + this.name);
		if (this.config.refreshRate < this.minRefreshRate) {
			this.config.refreshRate = this.minRefreshRate;
			Log.error("WARNING: refreshRate was configured to be lower than the minimum 30 seconds. MMM-HabiticaStats will use 30 second refresh rate instead.");
		}
		this.fetchUserData();

	},

	fetchUserData: function() {
		// If you're doing a significant fork of MMM-HabiticaStats you might want to change the xclient header here. For more info on the xclient header: https://habitica.fandom.com/wiki/Guidance_for_Comrades#X-Client_Header
		let xclient = "6c2c57d5-67c3-4edf-9a74-2d6d70aa4c56-MMM-HabiticaStats-v1.3.0";
		apiManager = new HabiticaAPIManager(xclient);
		// it looks like we're fetching the content data anew each time, but the
		// browser should have this cached so unless Habitica has updated the
		// content on their end this should end up being a local fetch to cache
		// after the first time
		apiManager.fetchContentData()
		.then(() => {
			return apiManager.fetchUserWithTasks(this.config.userID, this.config.APIToken);
		})
		.then((user) => {
			this.user = user;
			this.updateDom();
			this.scheduleUpdate();
		})
		.catch((fetchError) => {
			var error;
			try {
				error = JSON.parse(fetchError);
			} catch(e) {
				if (fetchError == "") {
					// internet is probably out... might as well try scheduling the next
					// update in case it comes back.
					this.scheduleUpdate();
					fetchError = "Could not connect to the Habitica server.";
				};
				// most of these other errors mean a badly reformatted request and
				// we shouldn't try to schedule new jobs -- user needs to correct the
				// error on their end first
				error = {error: fetchError};
			}
			this.fetchError = error;
			console.error(error);
			this.updateDom();
		});
	},

	scheduleUpdate: function() {
		this.updateTimer = setTimeout(() => {
				this.fetchUserData();
			},
			this.config.refreshRate
		);
	},

	suspend: function() {
		Log.info("Suspending MMM-HabiticaStats...");
		clearTimeout(this.updateTimer);
	},

	resume: function() {
		Log.info("Resuming MMM-HabiticaStats...");
		this.fetchUserData();
	},
});
